org.eclipse.jdt.ui/core refactoring/org/eclipse/jdt/internal/corext/refactoring/util/ChangedValueChecker.java

```diff
@@ -0,0 +1,582 @@
+				SimpleName sn= (SimpleName) node;
+				IBinding resolveBinding= sn.resolveBinding();
+				if (resolveBinding != null && resolveBinding.getJavaElement() != null) {
+					if (resolveBinding.getJavaElement().getElementType() != IJavaElement.LOCAL_VARIABLE
+							|| flag)
+						this.memberKey= resolveBinding.getKey();
+				}
+			} else if (node instanceof QualifiedName) {
+				QualifiedName qn= (QualifiedName) node;
+				SimpleName sn= qn.getName();
+				this.memberKey= sn.resolveBinding() != null ? sn.resolveBinding().getKey() : null;
+				if (qn.resolveBinding() != null && qn.resolveBinding().getModifiers() != Modifier.STATIC)
+					this.e= new Elem(qn.getQualifier(), flag);
+			} else if (node instanceof FieldAccess) {
+				FieldAccess fa= (FieldAccess) node;
+				SimpleName sn= fa.getName();
+				this.memberKey= sn.resolveBinding() != null ? sn.resolveBinding().getKey() : null;
+				ASTNode expr= getOriginalExpression(fa.getExpression());
+				if (expr instanceof MethodInvocation) {
+					if (flag) {
+						this.e= new Elem((MethodInvocation) expr);
+					} else {
+						this.memberKey= null;
+					}
+				} else
+					this.e= new Elem(expr, flag);
+			} else if (node instanceof MethodInvocation) {
+				MethodInvocation mi= (MethodInvocation) node;
+				if (flag) {
+					this.e= new Elem(mi);
+				} else {
+					this.memberKey= null;
+				}
+			}
+		}
+
+		public Elem(MethodInvocation expr) { // use String to represent the instances of MethodInvocation
+			super();
+			this.memberKey= expr.toString();
+		}
+
+		public Elem(ASTNode node, boolean flag, Elem e) {
+			if (node instanceof MethodInvocation && flag) {
+				MethodInvocation mi= (MethodInvocation) node;
+				this.e= new Elem(mi.getExpression(), flag);
+				this.memberKey= e;
+			}
+		}
+
+		@Override
+		public int hashCode() {
+			final int prime= 31;
+			int result= 1;
+			result= prime * result + getEnclosingInstance().hashCode();
+			result= prime * result + this.toString().hashCode();
+			return result;
+		}
+
+		@Override
+		public String toString() {
+			String str1= this.e == null ? "" : this.e.toString(); //$NON-NLS-1$
+			String str2= ""; //$NON-NLS-1$
+			if (this.memberKey instanceof String) {
+				str2= (String) this.memberKey;
+			} else if (this.memberKey instanceof Elem) {
+				str2= ((Elem) this.memberKey).toString();
+			}
+			if (!str1.equals("")) //$NON-NLS-1$
+				return str1 + str2;
+			else
+				return str2;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (this == obj) {
+				return true;
+			}
+			if (!(obj instanceof Elem)) {
+				return false;
+			}
+			Elem other= (Elem) obj;
+			if (!getEnclosingInstance().equals(other.getEnclosingInstance())) {
+				return false;
+			}
+			String str1= this.toString();
+			String str2= obj.toString();
+			return str1.equals(str2);
+		}
+
+		private ChangedValueChecker getEnclosingInstance() {
+			return ChangedValueChecker.this;
+		}
+	}
+
+	class PathVisitor extends ASTVisitor {
```

Comment:
Please add comment about what PathVisitor does as it is not obvious like ReadVisitor or UpdateVisitor.

Reply:
Looks good.  Thx.